File: 1297.maximum-number-of-occurrences-of-a-substring.py
#
# @lc app=leetcode id=1297 lang=python3
#
# [1297] Maximum Number of Occurrences of a Substring
#

# @lc code=start
from collections import Counter
class Solution:
    def maxFreq(self, s: str, maxLetters: int, minSize: int, maxSize: int) -> int:
        #! So the maxSize is just a distraction????
        #! The minSize is the key to the solution 
        # left = 0
        # hashmap = Counter()
        # for i in range(len(s)):
        #     #todo Enter the window
        #     if i < minSize -1:
        #         continue
        #     #todo Find the correct window size
        #     curr_str = s[left:i+1]

        #     #todo Check if the current window is valid
        #     if len(set(curr_str)) > maxLetters:
        #         left += 1
        #         continue
        #     #todo If the current window is valid, add it to the hashmap
        #     hashmap[curr_str] += 1
        #     left += 1
        # return max(hashmap.values()) if hashmap else 0

        hashmap_str = Counter()
        hashmap_letter = Counter()
        left = 0
        for right,c in enumerate(s):
            hashmap_letter[c] += 1
            if right < minSize - 1:
                continue
            if len(hashmap_letter) > maxLetters:
                hashmap_letter[s[left]] -= 1
                if hashmap_letter[s[left]] == 0:
                    del hashmap_letter[s[left]]
                left += 1
                continue
            hashmap_str[s[left:right+1]] += 1
            hashmap_letter[s[left]] -= 1
            if hashmap_letter[s[left]] == 0:
                del hashmap_letter[s[left]]
            left += 1
        return max(hashmap_str.values()) if hashmap_str else 0
# @lc code=end

File: 1297.maximum-number-of-occurrences-of-a-substring.java
/*
 * @lc app=leetcode id=1297 lang=java
 *
 * [1297] Maximum Number of Occurrences of a Substring
 */
import java.util.HashMap;
// @lc code=start
class Solution {
    public int maxFreq(String s, int maxLetters, int minSize, int maxSize) {
        char[] freq = new char[26];
        int n = s.length();
        HashMap<String, Integer> map = new HashMap<>();
        for(int i = 0; i < n; i++){
            freq[s.charAt(i) - 'a']++;
            
            if(i >= minSize -1){
                //count the unique letters in freq char array
                int count = 0;
                for(int j = 0; j < 26; j++){
                    if(freq[j] > 0){
                        count++;
                    }
                }
                if(count <= maxLetters){
                    String sub = s.substring(i - minSize + 1, i + 1);
                    map.merge(sub, 1, Integer::sum);
                }
                freq[s.charAt(i - minSize + 1) - 'a']--;
            }
        }
        return map.values().stream().max(Integer::compareTo).orElse(0);
    }
}
// @lc code=end

